Add PVSPEC spectral correction factor model

RDaxini opened the pull request:
- [X] Closes (partially addresses #1950 and #2065)
 - [X] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [X] Tests added
 - [X] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

[This model](https://ieeexplore.ieee.org/document/9300932) is one of the spectral correction functions suggested in the cited issues (#1950 and #2065).
Docs: https://pvlib-python--2072.org.readthedocs.build/en/2072/reference/generated/pvlib.spectrum.spectral_factor_pvspec.html

cwhanse commented on pvlib/tests/test_pelland.py:
```diff
@@ -0,0 +1,74 @@
+import numpy as np
+from warnings import warn
+
+def spectral_factor_pelland(airmass_absolute, clearness_index,
```
Add 'test_' at the front of each function you want pytest to run.

echedey-ls commented:
There are a few things that need to be done first:

1. Separate the implementation into an existing or new file. E.g.: [`pvlib/spectrum/mismatch.py`](https://github.com/pvlib/pvlib-python/blob/main/pvlib/spectrum/mismatch.py)
   (That's moving the current function you have to there, modifying the needed `__init__.py` files too; I think your current `test_pelland.py` file is not a test, it seems like an implementation of the model)
2. Create tests under `pvlib/tests/test_spectrum.py`. Prefix tests with `test_` as @cwhanse noted, so they can be discovered automatically by pytest. I don't know exactly the pytest command to run exactly one test by name, but you could run that file by itself there. Or keep that file temporarily.

In general, you may want to see some PRs that add new functions as a reference. For example, #2041 is simple IMO. There are many more of course, it's just I'm too selfish to look for other ones (/sarcasm)

I've been inspired by #2067 to do a post blog regarding my setup. I know I wrote too much, but in case it helps you, check it out [here](https://echedey-ls.github.io/posts/my-setup/).

cwhanse commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,128 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
     modifier = f_AM + f_AOD + f_PW  # Eq 5
     return modifier
+
+def test_spectral_factor_pelland(airmass_absolute, clearness_index,
```
@RDaxini sorry I wasn't clear. Use 'test_' as the first part of function names in the 'test_spectrum.py' fiel that pytest processes. Here, in mismatch.py, we don't put 'test_' in the function name.

RDaxini commented:
@echedey-ls @cwhanse Thank you for the help. I think I have updated the files with the correct names now (and also put the functions in the right places)
The tests that I have added into `test_spectrum.py` so far have passed, but are there any additional tests I need?
Something else I am unsure about is documentation. I added in text for the docs into `mismatch.py` but when I try to [view the build here](https://readthedocs.org/projects/pvlib-python/builds/24552940/) it looks just like the pvlib homepage rather than the text I included in `mismatch.py` so I wanted to ask how/where do I write/set up the documentation correctly?

echedey-ls commented:
Well done @RDaxini , it looks good!!

> how/where do I write/set up the documentation correctly

Yeah, documented functions aren't discovered automatically, you need to specify in a rST file the function. In general, these files are under `docs\sphinx\source\reference`. Specifically, for this model you would add to `docs\sphinx\source\reference\effects_on_pv_system_output\spectrum.rst`.

I will review in short some of the common typos we all make in the docstrings.

RDaxini commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,128 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    Parameters
+    ----------
+    airmass_absolute : numeric
+        absolute (pressure-adjusted) airmass. [unitless]
```
They use the Kasten and Young model, but I don't think that's relevant here. This part just defines variable names, and the user is free to derive values for those variables as they wish. I think, if the user wants to know specifics of any aspect of the methodology used by these authors, they could consult the paper.

RDaxini replied:
@echedey-ls actually I think I can include a brief summary of info such as this, but I'll add it into the intro above the parameters section rather than the parameters section itself

cwhanse commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,128 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
     modifier = f_AM + f_AOD + f_PW  # Eq 5
     return modifier
+
+def spectral_factor_pelland(airmass_absolute, clearness_index,
```
`clearness_index` needs to be `clearsky_index`. Clearness index is GHI / extraterrestrial; clearsky is GHI / clearsky model GHI.

Strictly speaking, the parameters for this model are tied to the ESRA clear-sky model (Rigollier, Bauer, Wald 2000, in the references of Pelland et al.) But that model looks complicated. For this addition of `spectral_factor_pelland`, a Note about the parameters would be helpful.

RDaxini replied:
@cwhanse Glad you picked up on this; an oversight on my part. I have corrected and added in a notes section with further details on the parameters.

RDaxini commented:
Some formatting errors in the docs since the last commit. Will leave it as it is for now and get back onto it tomorrow to address those.

echedey-ls commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,126 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    airmass and clear sky index using the Pelland model, which takes the
+    following form:
+
+    .. math::
```
One of the mentors commented on the last meeting that equations are usually on the notes section. Can this be moved there?

RDaxini replied:
@echedey-ls Yes I was going to ask about this as I thought they did but then couldn't remember when updating the script. I will make the change.

echedey-ls reviewed (commented):
I really like how this PR looks :)
Few comments regarding docs.

RDaxini commented:
I have now taken this PR out of draft mode and marked it ready for a full review

RDaxini commented:
Bit unsure about the last test failure. The commit before the last one had 1 failure due to a single trailing white space, which I fixed in the last commit. That test now passes, but a different test that passed previously now fails. I do not understand the error message, but I don't think it can be caused by the deletion of that trailing white space...
Could someone please help by explaining that error to me? Thank you

kandersolar commented:
The "Upload coverage to codecov" step of the test jobs is flaky and fails randomly.  Whether it fails has nothing to do with the contents of your PR, so feel free to ignore those failures.  We will re-trigger manually them whenever they fail.

RDaxini commented:
@kandersolar Got it, thanks.
So, what are the next steps now?

kandersolar commented:
Yep, next step is to wait for additional reviews to appear :)  Eventually someone will be satisfied that the review process is sufficiently complete, "approve" the PR, and merge it into pvlib's main branch.

kandersolar commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,127 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+       factor-ozone-watervapor-and-angstroembeta`_.
+    """
+
+# =============================================================================
+#      --- Default coefficients ---
+# =============================================================================
```
We don't use this comment style much (or at all?) in pvlib -- prefer simple `# here is the comment text` style where necessary.  In this case I think the code is sufficiently self-explanatory that I would just remove these three "header" comments altogether.

kandersolar commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,127 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+        0.9791, -0.03904, 0.03096)
+    _coefficients['asi'] = (
+        1.051, -0.1033, 0.009838)
+    _coefficients['polysi'] = _coefficients['multisi']
+    _coefficients['xsi'] = _coefficients['monosi']
```
These strings aren't advertised in the docstring as being acceptable.  I suggest removing them here, or if it's important to keep them for some reason (why?) then the docstring should list them as aliases.

kandersolar commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,127 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    elif module_type is None and coefficients is None:
+        raise ValueError('No valid input provided, both module_type and ' +
+                         'coefficients are None. module_type can be one of ' +
+                         'poly-Si, monosi, fs-2, fs-4, cigs, or asi')
```
```suggestion
                         ", ".join(_coefficients.keys()))
```

This way we don't have to list the keys in more than one place -- it makes it impossible to forget to list one, or forget to update the name here if we change it in the dictionary, etc.

RDaxini replied:
Ah this is a very good point

RDaxini replied:
Ah this is a good point

kandersolar commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,127 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    else:
+        raise ValueError('Cannot resolve input, must supply only one of ' +
+                         'module_type and coefficients. module_type can be ' +
+                         'one of poly-Si, monosi, fs-2, fs-4, cigs, or asi')
```
Same `", ".join(...)` trick here

kandersolar commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,127 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+        One of the following PV technology strings from [1]_:
+
+        * ``'fs4-1'`` - First Solar series 4-1 and earlier CdTe module.
+        * ``'fs4-2'`` - First Solar 4-2 and earlier CdTe module.
```
```suggestion
        * ``'fs4-2'`` - First Solar 4-2 and later CdTe module.
```

kandersolar commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,127 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
     return modifier
+
+
+def spectral_factor_pelland(airmass_absolute, clearsky_index,
```
I think we should consider calling this function `spectral_factor_pvspec`, as "PVSPEC" is the name given by the model's authors.

kandersolar commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,127 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+
+    where M is the spectral mismatch factor, and :math:`a_1, a_2, a_3` are
+    module-specific coefficients. In the PVSPEC model publication, absolute air
+    mass is estimated using the Kasten and Young model [2]_. The clear sky
```
```suggestion
    mass is estimated starting from the Kasten and Young relative air mass [2]_. The clear sky
```

I think many users are not savvy to the difference between absolute and relative air mass, so we should be careful to make it clear that K&Y on its own is not enough.  Feel free to suggest a clearer alternative :)

RDaxini replied:
Good point

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+])
+def test_spectral_factor_pelland(module_type, expected):
+    ams = np.array([1.0, 1.5, 2.0])
+    kcs = np.array([0.4, 0.6, 0.8])
```
This might be silly, but can we include a set of conditions that results in M < 1 for at least one of the modules?

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+
+def test_spectral_factor_pelland_supplied():
+    # use the multisi coeffs
+    coeffs = (
+        0.9847, -0.05237, 0.03034)
```
```suggestion
    coeffs = (0.9847, -0.05237, 0.03034)
```

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+    kcs = np.array([0.4, 0.6, 0.8])
+    out = spectrum.spectral_factor_pelland(ams, kcs,
+                                           module_type=module_type)
+    assert np.allclose(expected, out, atol=1e-3)
```
Seems like we can do better than `1e-3` here.  Looking at the expected values, how about `1e-7` or `1e-8`?

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+        0.9847, -0.05237, 0.03034)
+    out = spectrum.spectral_factor_pelland(1.5, 0.8, coefficients=coeffs)
+    expected = 1.00860641
+    assert_allclose(out, expected, atol=1e-3)
```
Same here

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+    # Error when specifying both module_type and coefficients
+    coeffs = (
+        0.9847, -0.05237, 0.03034)
+    with pytest.raises(ValueError):
```
```suggestion
    with pytest.raises(ValueError, match='supply only one of'):
```

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+
+def test_spectral_factor_pelland_supplied_redundant():
+    # Error when specifying both module_type and coefficients
+    coeffs = (
+        0.9847, -0.05237, 0.03034)
```
```suggestion
    coeffs = (0.9847, -0.05237, 0.03034)
```

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+
+def test_spectral_factor_pelland_supplied_ambiguous():
+    # Error when specifying neither module_type nor coefficients
+    with pytest.raises(ValueError):
```
```suggestion
    with pytest.raises(ValueError, match='No valid input provided'):
```

kandersolar commented on pvlib/tests/test_spectrum.py:
```diff
@@ -315,3 +315,44 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+    ('fs-4', np.array([1.05234725, 1.037771, 1.0275516])),
+    ('multisi', np.array([1.03310403, 1.02391703, 1.01744833])),
+    ('monosi', np.array([1.03225083, 1.02335353, 1.01708734])),
+    ('cigs', np.array([1.01475834, 1.01143927, 1.00909094])),
```
Can we include a test where the inputs and outputs are pandas Series?

You could re-use the same expected value variables, just convert them to Series and re-run the function like this example: https://github.com/pvlib/pvlib-python/blob/main/pvlib/tests/test_iam.py#L60-L63

kandersolar reviewed (commented):
This is looking great @RDaxini!

RDaxini commented:
Think I have now made all revisions suggested by @kandersolar

IoannisSifnaios commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,111 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    the spectrum, and thus spectral mismatch factor estimation. Another
+    motivation is to develop a simple parameterisation compared with existing
+    models. Model coefficients are derived using spectral irradiance and other
+    meteorological data from eight locations. These coefficients for seven
+    modules, available here via the ``module_type`` parameter, as well as more
+    details on the model, can be found in [1]_.
```
```suggestion
    meteorological data from eight locations. The coefficients for six module
    types are available via the ``module_type`` parameter. More
    details on the model, can be found in [1]_.
```

IoannisSifnaios commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,111 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+
+    .. math::
+
+       M = a_1 K_c^{a_2} AM_a^{a_3},
```
```suggestion
       M = a_1 k_c^{a_2} AM^{a_3},
```

IoannisSifnaios replied:
Just to stick to the convention followed in the paper

RDaxini replied:
@IoannisSifnaios I will change the kc,
For AM, is it better to stick with the paper or pvlib convention? The paper just uses AM for absolute airmass, but AM_a is used for other spectral corrections in pvlib so I thought it would be easier for users if I kept the definitions the same as with the other spectral corrections in pvlib.
Thanks for reviewing btw

IoannisSifnaios replied:
@kandersolar, thoughts on this?

kandersolar replied:
Consistency within pvlib is the priority.  When that means using terminology that doesn't match the reference, it can be helpful to note the difference as a comment in the code or in the docstring as appropriate.

IoannisSifnaios commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,111 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+
+       M = a_1 K_c^{a_2} AM_a^{a_3},
+
+    where M is the spectral mismatch factor, and :math:`a_1, a_2, a_3` are
```
```suggestion
    where :math:`M` is the spectral mismatch factor, :math:`k_c` is the clear sky index, :math:`AM` is the absolute air mass, and :math:`a_1, a_2, a_3` are
```

IoannisSifnaios replied:
I suggest you also add Table III (that has all the "a" coefficients) as a table in the "Notes" section.

IoannisSifnaios commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,111 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+
+    References
+    ----------
+    .. [1] Pelland, S., Beswick, C., Thevenard, D., Côté, A., Pai, A. and
+       Poissant, Y., 2020, June. Development and testing of the PVSPEC model of
+       photovoltaic spectral mismatch factor. In 2020 47th IEEE Photovoltaic
+       Specialists Conference (PVSC) (pp. 1258-1264). IEEE.
+       :doi:`10.1109/PVSC45281.2020.9300932`
```
Something weird happens with this reference when you build the docs... maybe @kandersolar or @AdamRJensen can help?

RDaxini replied:
What is the weird thing that happens?

IoannisSifnaios replied:
This is how it looks when you open the readthedocs:

![image](https://github.com/pvlib/pvlib-python/assets/88548539/f5316c8c-4567-4f30-bf88-5b45237524d0)

RDaxini replied:
You mean the (1,2) ? That is just because there are two references to this paper. If you click the 1 or the 2 in the brackets, it takes you to the first or the second reference to the paper in the text above.

kandersolar replied:
Yep, that's just how the docs output looks.  We could look into changing it (I don't see much value in the (1, 2) and it confuses everyone at first), but that should definitely be a separate PR.  Nothing to fix here :)

RDaxini replied:
Yeah can't lie it threw me too at first, then I realised it was just a hyperlink to the in-text references

IoannisSifnaios commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,111 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    the ESRA model [3]_ to estimate the clear sky GHI with monthly Linke
+    turbidity values from [4]_ as inputs.
+
+    References
+    ----------
+    .. [1] Pelland, S., Beswick, C., Thevenard, D., Côté, A., Pai, A. and
+       Poissant, Y., 2020, June. Development and testing of the PVSPEC model of
+       photovoltaic spectral mismatch factor. In 2020 47th IEEE Photovoltaic
+       Specialists Conference (PVSC) (pp. 1258-1264). IEEE.
+       :doi:`10.1109/PVSC45281.2020.9300932`
+    .. [2] Kasten, F. and Young, A.T., 1989. Revised optical air mass tables
+       and approximation formula. Applied Optics, 28(22), pp.4735-4738.
+       :doi:`https://doi.org/10.1364/AO.28.004735`
+    .. [3] Rigollier, C., Bauer, O. and Wald, L., 2000. On the clear sky model
+       of the ESRA—European Solar Radiation Atlas—with respect to the Heliosat
+       method. Solar energy, 68(1), pp.33-48.
+       :doi:`https://doi.org/10.1016/S0038-092X(99)00055-9`
+    .. [4] SoDa website monthly Linke turbidity values:
+       `http://www.sodapro.com/gl/web-services/atmosphere/linke-turbidity-
+       factor-ozone-watervapor-and-angstroembeta`_.
```
Do we have an "official" convention on how we do references? In most places, I see that when the authors are more than 2, we just write the first author and then et al. 
Same question on how to report the journal issue and pages.
@kandersolar, thoughts?

kandersolar replied:
No, we don't enforce or even prefer a specific format.  My two cents: authors, title, year, journal, and DOI are the necessary elements.  "et al." is allowed but not always used.  I am personally okay with these references as currently formatted.

IoannisSifnaios commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,111 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    References
+    ----------
+    .. [1] Pelland, S., Beswick, C., Thevenard, D., Côté, A., Pai, A. and
+       Poissant, Y., 2020, June. Development and testing of the PVSPEC model of
```
```suggestion
       Poissant, Y., 2020. Development and testing of the PVSPEC model of
```

IoannisSifnaios reviewed (commented):
Some ideas from my side

echedey-ls commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,112 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+       :doi:`10.1109/PVSC45281.2020.9300932`
+    .. [2] Kasten, F. and Young, A.T., 1989. Revised optical air mass tables
+       and approximation formula. Applied Optics, 28(22), pp.4735-4738.
+       :doi:`https://doi.org/10.1364/AO.28.004735`
```
```suggestion
       :doi:`10.1364/AO.28.004735`
```

echedey-ls commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,112 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+    .. [3] Rigollier, C., Bauer, O. and Wald, L., 2000. On the clear sky model
+       of the ESRA—European Solar Radiation Atlas—with respect to the Heliosat
+       method. Solar energy, 68(1), pp.33-48.
+       :doi:`https://doi.org/10.1016/S0038-092X(99)00055-9`
```
```suggestion
       :doi:`10.1016/S0038-092X(99)00055-9`
```

echedey-ls commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,112 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+
+    .. math::
+
+       M = a_1 k_c^{a_2} AM_a^{a_3},
```
```suggestion
       M = a_1 k_c^{a_2} AM_a^{a_3}
```

RDaxini replied:
Nooo... I am a proponent of punctuating equations as they are part of the sentence even when in display math mode 😄

echedey-ls commented on pvlib/spectrum/mismatch.py:
```diff
@@ -571,3 +571,112 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+       method. Solar energy, 68(1), pp.33-48.
+       :doi:`https://doi.org/10.1016/S0038-092X(99)00055-9`
+    .. [4] SoDa website monthly Linke turbidity values:
+       `http://www.sodapro.com/gl/web-services/atmosphere/linke-turbidity-
+       factor-ozone-watervapor-and-angstroembeta`_.
+    """
```
```suggestion
       `http://www.sodapro.com/gl/web-services/atmosphere/linke-turbidity-factor-ozone-watervapor-and-angstroembeta`_.
    """  # noqa: E501
```
Let's see how this looks. If it doesn't work, you may need to remove backticks and underscore.

In case of doubt: https://bashtage.github.io/sphinx-material/rst-cheatsheet/rst-cheatsheet.html
or https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages

echedey-ls reviewed (approved):
LGTM. I left comments about broken links.

RDaxini commented:
>  left comments about broken links.
@echedey-ls I am sure I changed these according to your suggestion but I did get mixed up between my local version and the version on github at one point (committing suggested changes online and making more changes locally) and these changes must have been lost. Thanks for re-highlighting this

echedey-ls commented:
Don't worry @RDaxini , we've all been there.

Btw, can you double check the "sodapro" link is what you intended? http://www.sodapro.com/gl/web-services/atmosphere/linke-turbidity-factor-ozone-watervapor-and-angstroembeta

It doesn't work (404). Ah, and Sodapro looks like a sandblasting company?

cwhanse commented on pvlib/spectrum/mismatch.py:
```diff
@@ -583,6 +583,114 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+
+        * ``'fs4-1'`` - First Solar series 4-1 and earlier CdTe module.
+        * ``'fs4-2'`` - First Solar 4-2 and later CdTe module.
+        * ``'monosi'``, - anonymous sc-si module.
```
```suggestion
        * ``'monosi'``, - anonymous monocrystalline Si module.
```

cwhanse commented on pvlib/spectrum/mismatch.py:
```diff
@@ -583,6 +583,114 @@ def spectral_factor_caballero(precipitable_water, airmass_absolute, aod500,
+        * ``'fs4-1'`` - First Solar series 4-1 and earlier CdTe module.
+        * ``'fs4-2'`` - First Solar 4-2 and later CdTe module.
+        * ``'monosi'``, - anonymous sc-si module.
+        * ``'multisi'``, - anonymous mc-si- module.
```
```suggestion
        * ``'multisi'``, - anonymous multicrystalline Si module.
```

cwhanse commented on pvlib/tests/test_spectrum.py:
```diff
@@ -317,6 +318,62 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+    kcs = pd.Series([0.4, 0.6, 0.8, 1.4])
+    out = spectrum.spectral_factor_pvspec(ams, kcs,
+                                          module_type=module_type)
+    out = pd.Series(out)
```
If `spectral_factor_pvspec` gets `Series` input I would expect `Series` output.

cwhanse replied:
Yes I think that line should be removed from the test, and the test should check that out is a Series

RDaxini replied:
Understood. Commit [a772a94](https://github.com/pvlib/pvlib-python/pull/2072/commits/a772a94fbeb3d6e507caa2147e87c49a20808556) should resolve this

cwhanse reviewed (commented):
Editorial mostly, I think that matching output type to input type has slso been raised by others.

echedey-ls commented on pvlib/tests/test_spectrum.py:
```diff
@@ -317,6 +318,62 @@ def test_spectral_factor_caballero_supplied_ambiguous():
+    kcs = pd.Series([0.4, 0.6, 0.8, 1.4])
+    out = spectrum.spectral_factor_pvspec(ams, kcs,
+                                          module_type=module_type)
+    if isinstance(out, pd.Series):
+        assert np.allclose(expected, out, atol=1e-8)
```
That will only execute the values assertion if the output type is a `pd.Series`. It won't check that the output is actually that object.

```suggestion
    assert isinstance(out, pd.Series)
    assert np.allclose(expected, out, atol=1e-8)
```

RDaxini replied:
ohh... my (original) thinking was that if the output is not `pd.Series` then because the values assertion doesn't execute, the test fails. 
I have committed your suggestion. Thanks

echedey-ls replied:
Don't worry. Just in case, a test only fails when an exception of any kind is raised (and it is not inside an exception block, like a `pytest.raises` to check that an exception raises -> that block raises an exception if no exception is raised).

echedey-ls commented:
@kandersolar @AdamRJensen or @cwhanse can you approve the workflows on this PR?

kandersolar commented:
@RDaxini one last thing is needed: a "what's new" entry for v0.11.0.  Can you make a new bullet point under `Enhancements` in [v0.11.0.rst](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/whatsnew/v0.11.0.rst), and list yourself in the Contributors list at the bottom of the file?

kandersolar commented on docs/sphinx/source/whatsnew/v0.11.0.rst:
```diff
@@ -32,7 +32,9 @@ Enhancements
   :py:func:`pvlib.spectrum.sr_to_qe` and :py:func:`pvlib.spectrum.qe_to_sr`
   respectively. (:issue:`2040`, :pull:`2041`)
-
+* Add function :py:func`spectrum.spectral_factor_pvspec`, which calculates the
```
```suggestion
* Add function :py:func:`pvlib.spectrum.spectral_factor_pvspec`, which calculates the
```

kandersolar reviewed (approved):
LGTM with one minor syntax edit.  Thanks @RDaxini!

RDaxini commented:
Thanks for all the reviews @echedey-ls @cwhanse @IoannisSifnaios @kandersolar
